yarn-project/simulator/src/avm/opcodes/multi_scalar_mul.test.ts

```diff
@@ -0,0 +1,142 @@
+    // Points are stored as [x1, y1, inf1, x2, y2, inf2, ...] where the types are [Field, Field, Uint8, Field, Field, Uint8, ...]
+    for (let i = 0; i < points.length; i++) {
+      const flatPointsOffset = pointsOffset + 2 * i; // 2 since we only have x and y
+      const memoryOffset = pointsOffset + 3 * i; // 3 since we store x, y, inf
+      context.machineState.memory.set(memoryOffset, storedPoints[flatPointsOffset]);
+      context.machineState.memory.set(memoryOffset + 1, storedPoints[flatPointsOffset + 1]);
+      context.machineState.memory.set(memoryOffset + 2, new Uint8(points[i].isInfPoint() ? 1 : 0));
+    }
+    // Store scalars
+    const scalarsOffset = pointsOffset + pointsReadLength;
+    context.machineState.memory.setSlice(scalarsOffset, storedScalars);
+    // Store length of points to read
+    const pointsLengthOffset = scalarsOffset + scalarsLength;
+    context.machineState.memory.set(pointsLengthOffset, new Uint32(pointsReadLength));
+    const outputOffset = pointsLengthOffset + 1;
+
+    await new MultiScalarMul(indirect, pointsOffset, scalarsOffset, outputOffset, pointsLengthOffset).execute(context);
+
+    const result = context.machineState.memory.getSlice(outputOffset, 3);
+
+    // We write it out explicitly here
+    let expectedResult = grumpkin.mul(points[0], scalars[0]);
+    expectedResult = grumpkin.add(expectedResult, grumpkin.mul(points[1], scalars[1]));
+    expectedResult = grumpkin.add(expectedResult, grumpkin.mul(points[2], scalars[2]));
+
+    expect(result.map(r => r.toFr())).toEqual([expectedResult.x, expectedResult.y, new Fr(0n)]);
+  });
+
+  it('Should perform msm correctly - indirect', async () => {
+    const indirect = 7;
+    const grumpkin = new Grumpkin();
+    // We need to ensure points are actually on curve, so we just use the generator
+    // In future we could use a random point, for now we create an array of [G, 2G, 3G]
+    const points = Array.from({ length: 3 }, (_, i) => grumpkin.mul(grumpkin.generator(), new Fq(i + 1)));
+
+    // Pick some big scalars to test the edge cases
+    const scalars = [new Fq(Fq.MODULUS - 1n), new Fq(Fq.MODULUS - 2n), new Fq(1n)];
+    const pointsReadLength = points.length * 3; // multiplied by 3 since we will store them as triplet in avm memory
+    const scalarsLength = scalars.length * 2; // multiplied by 2 since we will store them as lo and hi limbs in avm memory
+    // Transform the points and scalars into the format that we will write to memory
+    // We just store the x and y coordinates here, and handle the infinities when we write to memory
+    const storedPoints: Field[] = points.flatMap(p => [new Field(p.x), new Field(p.y)]);
+    const storedScalars: Field[] = scalars.flatMap(s => [new Field(s.low), new Field(s.high)]);
+
+    const pointsOffset = 0;
+    // Store points...awkwardly (This would be simpler if ts handled the infinities in the Point struct)
```

Comment:
Something optional that you could use to modify the Point class without modifying too much

https://www.typescriptlang.org/play/?#code/MYGwhgzhAEAKD2BLAdgF2gbwFDWseyEqATgK7CrzEAUADqQEYiLDQAeAXMqQLYMCmxADTR6TFtACeXXgOIBKTAF8sOUY2asA5v3QoAZtUXZcuYrtLFk0VAAtEEAHRtoAXlfQADAG41KtWKaNvAAYoj8IAAmEADqiHYAksiGxmpmFlbQANp2Ds4iuU6SBfZOBgC6vrgq-viE6LRu0Mj8AO5wSGjUniIALPK+dRDwIPyOIPBadI4GA1hDI2MTU7SOlGER0XGJyUZzWEA